contracts/limitOrder/LimitOrderFeeVault.sol

```diff
@@ -0,0 +1,101 @@
+import { SignedSafeMathUpgradeable } from "@openzeppelin/contracts-upgradeable/math/SignedSafeMathUpgradeable.sol";
+import { PerpMath } from "@perp/curie-contract/contracts/lib/PerpMath.sol";
+import { ILimitOrderFeeVault } from "../interface/ILimitOrderFeeVault.sol";
+import { LimitOrderFeeVaultStorageV1 } from "../storage/LimitOrderFeeVaultStorage.sol";
+import { OwnerPausable } from "../base/OwnerPausable.sol";
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+// solhint-disable-next-line max-line-length
+import { SafeERC20Upgradeable, IERC20Upgradeable } from "@openzeppelin/contracts-upgradeable/token/ERC20/SafeERC20Upgradeable.sol";
+
+contract LimitOrderFeeVault is
+    ILimitOrderFeeVault,
+    BlockContext,
+    ReentrancyGuardUpgradeable,
+    OwnerPausable,
+    LimitOrderFeeVaultStorageV1
+{
+    using AddressUpgradeable for address;
+    using PerpMath for int256;
+    using PerpMath for uint256;
+    using SignedSafeMathUpgradeable for int256;
+
+    modifier onlyLimitOrderBook() {
+        // LOFV_SMBLOB: Sender Must Be LimitOrderBook
+        require(_msgSender() == limitOrderBook, "LOFV_SMBLOB");
+        _;
+    }
+
+    function initialize(address rewardTokenArg, uint256 feeAmountArg) external initializer {
+        __OwnerPausable_init();
+        __ReentrancyGuard_init();
+
+        // LOFV_RTINC: RewardToken Is Not a Contract
+        require(rewardTokenArg.isContract(), "LOFV_RTINC");
+        rewardToken = rewardTokenArg;
+
+        // LOFV_FAMBGT0: FeeAmount Must Be Greater Than 0
+        require(feeAmountArg > 0, "LOFV_FAMBGT0");
+        feeAmount = feeAmountArg;
+    }
+
+    function setRewardToken(address rewardTokenArg) external onlyOwner {
+        // LOFV_RTINC: RewardToken Is Not a Contract
+        require(rewardTokenArg.isContract(), "LOFV_RTINC");
+        rewardToken = rewardTokenArg;
+        emit RewardTokenChanged(rewardTokenArg);
+    }
+
+    function setLimitOrderBook(address limitOrderBookArg) external onlyOwner {
+        // LOFV_LOBINC: LimitOrderBook Is Not a Contract
+        require(limitOrderBookArg.isContract(), "LOFV_LOBINC");
+        limitOrderBook = limitOrderBookArg;
+        emit LimitOrderBookChanged(limitOrderBookArg);
+    }
+
+    function setFeeAmount(uint256 feeAmountArg) external onlyOwner {
+        // LOFV_FAMBGT0: FeeAmount Must Be Greater Than 0
+        require(feeAmountArg > 0, "LOFV_FAMBGT0");
+        feeAmount = feeAmountArg;
+        emit FeeAmountChanged(feeAmountArg);
+    }
+
+    function disburse(address keeper, uint256 orderValue)
+        external
+        override
+        onlyLimitOrderBook
+        nonReentrant
+        returns (uint256)
+    {
+        // TODO: be aware of decimal issue when we use different reward token
+
+        // LOFV_NEBTD: not enough balance to disburse
+        require(IERC20Upgradeable(rewardToken).balanceOf(address(this)) >= feeAmount, "LOFV_NEBTD");
```

Comment:
question, do we really want to revert here? If so, it'll cause trader's order being reverted and
block keepers like us who are willing to execute the orders without feedback.

Reply:
I guess we need to setup monitor(like grafana or oncall duty) to check if this balance is enough to disburse.

Besides, if we run out of perp token in this LimitOrderFeeVault, we still make this order filled but won't send any reward to users. Later we found this situation happened, we need to write an off-chain script to send reward to those keepers, which could take us more effort to handle it).

Reply:
Besides, if one filled order's keeper got keeper reward, but another filled order's keeper doesn't, it would be confused since both transactions are successful.